machetli/search.py

```diff
@@ -86,18 +88,62 @@ def search(initial_state, successor_generator, evaluator_path, environment=None)
     current_state = initial_state
     while True:
         successors = successor_generator.get_successors(current_state)
-        for batch in batched(successors, environment.batch_size):
-            try:
-                environment.submit(batch, evaluator_path)
-                environment.wait_until_finished()
-                successor = environment.get_improving_successor()
-            except (SubmissionError, TaskError, PollingError):
-                # FIXME: this is not proper error handling yet.
-                successor = None
-            if successor:
-                logging.info(successor.change_msg)
-                current_state = successor.state
-                break
+        try:
+            improving_state, message = _get_improving_successor(
+                evaluator_path, successors, environment)
+        except SubmissionError as e:
+            logging.critical(f"Terminating search because job submission for successor evaluation failed:\n{e}")
+        except PollingError as e:
+            logging.critical(f"Terminating search because querying the status of a submitted successor evaluation failed:\n{e}")
+
+        if message:
+            logging.info(message)
+        if improving_state:
+            current_state = improving_state
         else:
-            logging.info("No improving successor found, terminating search.")
             return current_state
+
+
+def _get_improving_successor(evaluator_path, successors, environment):
+    report_if_no_improvement = set()
+    def on_task_completed(job_id, task):
+        if environment.allow_nondeterministic_successor_choice:
+            # If an evaluation reports a non-improving successor,
+            # deterministic mode considers the next successor. In all
+            # other cases (improving successor, evaluation did not
+            # finish correctly), there is no need to evaluate later
+            # successors.
```

Comment:
The comment above seems a bit redundant with the code expressing pretty much the same.